File: JAVA/Binary Tree Level Order Traversal II.java
public class Solution {
    public ArrayList<ArrayList<Integer>> levelOrderBottom(TreeNode root) {
        ArrayList<ArrayList<Integer>> ans = new ArrayList<ArrayList<Integer>>();
        Queue<TreeNode> queue = new LinkedList<TreeNode>();
        queue.offer(root);
        int lastLevel = 1;
        int thisLevel = 0;
        ArrayList<Integer> tmp = new ArrayList<Integer>();
        while (queue.size() != 0)
        {
            TreeNode n = queue.poll();
            lastLevel--;
            if (n != null)
            {
                tmp.add(n.val); 
                queue.offer(n.left);
                queue.offer(n.right);
                thisLevel += 2;
            }
            if (lastLevel == 0)
            {
                lastLevel = thisLevel;
                thisLevel = 0;
                if (tmp.size() != 0)
                {
                    ans.add(tmp);
                }
                tmp = new ArrayList<Integer>();
            }
        }
        int len = ans.size();
        for (int i = 0; i < len/2; i++)
        {
            tmp = ans.get(i);
            ans.set(i, ans.get(len - i - 1));
            ans.set(len - i - 1, tmp);
        }
 
        return ans;
    }
}

File: Binary Tree Level Order Traversal II.cpp
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<vector<int> > levelOrderBottom(TreeNode *root) {
        vector<vector<int> > result;
        if (root == NULL) return result;
        queue<TreeNode *> que;
        que.push(root);
        int thisLevel = 1;
        int nextLevel = 0;
        vector<int> level;
        while (not que.empty()) {
            TreeNode *node = que.front();
            que.pop();
            level.push_back(node->val);
            if (node->left != NULL) {
                que.push(node->left);
                nextLevel++;
            }
            if (node->right != NULL) {
                que.push(node->right);
                nextLevel++;
            }
            thisLevel--;
            if (thisLevel == 0) {
                thisLevel = nextLevel;
                nextLevel = 0;
                result.push_back(level);
                level.clear();
            }
        }
        reverse(result.begin(), result.end());
        return result;
    }
};
